src/OrleansGetEventStore/GetEventStoreProvider.cs

```diff
@@ -0,0 +1,304 @@
+                    ToVersion = versionThatExists,
+                    Events = emptyList
+                };
+            }
+
+            // general case: load segment in batches
+            var sliceStart = fromVersion;
+            StreamEventsSlice currentSlice = null;
+            var events = new List<KeyValuePair<Guid,object>>();
+            do
+            {
+                var readPageSize = Math.Min(pageSize, (toVersion ?? int.MaxValue) - sliceStart);
+
+                if (readPageSize == 0)
+                    break;
+
+                currentSlice = await this.Connection.ReadStreamEventsForwardAsync(streamName, sliceStart, readPageSize, true);
+
+                if (currentSlice.Status == SliceReadStatus.StreamNotFound
+                    || currentSlice.Status == SliceReadStatus.StreamDeleted)
+
+                    // stream does not exist... return an empty segment
+                    return new EventStreamSegment()
+                    {
+                        StreamName = streamName,
+                        FromVersion = 0,
+                        ToVersion = 0,
+                        Events = emptyList
+                    };
+
+                if (fromVersion > currentSlice.LastEventNumber)
+                {
+                    // the fromVersion is larger than last event in the slice; 
+                    // this means the query targeted a non-existing segment. Return empty segment.
+                    return new EventStreamSegment()
+                    {
+                        StreamName = streamName,
+                        FromVersion = currentSlice.LastEventNumber + 1,
+                        ToVersion = currentSlice.LastEventNumber + 1,
+                        Events = emptyList
+                    };
+                }
+
+                foreach (var evnt in currentSlice.Events)
+                      events.Add(new KeyValuePair<Guid,object>(evnt.Event.EventId, DeserializeEvent(evnt.Event)));
+
+                sliceStart = currentSlice.NextEventNumber;
+
+            } while (!currentSlice.IsEndOfStream);
+
+            return new EventStreamSegment()
+            {
+                StreamName = streamName,
+                FromVersion = fromVersion,
+                ToVersion = fromVersion + events.Count,
+                Events = events
+            };
+        }
+
+        public async Task<bool> Delete(string streamName, int? expectedVersion)
+        {
+            try
+            {
+                var versionParameter = expectedVersion.HasValue ? expectedVersion.Value - 1 : ExpectedVersion.Any;
+                var result = await this.Connection.DeleteStreamAsync(streamName, versionParameter, true);
+                return true;
+            }
+            catch (WrongExpectedVersionException)
+            {
+                return false;
+            }
+            catch (StreamDeletedException)
+            {
+                return false;
```

Comment:
we are not giving idempotent semantics for deletion... that is a bad idea here as it means we would have to remember the position of the deleted stream. 

Once deleted, the thing is gone. If you are not sure if you succeeded, you can go check the version - it is zero if the deletion succeeded.